src/core/pwm.c

```diff
@@ -0,0 +1,161 @@
+#include <pic32_pwm.h>
+#include <stdio.h>
+#include <letmecreate/core/pwm.h>
+#include <letmecreate/core/common.h>
+
+/* Default period is 333333ns, in other words 3KHz */
+#define DEFAULT_PERIOD          (333333)
+
+int pwm_init(uint8_t mikrobus_index)
+{
+    if (check_valid_mikrobus(mikrobus_index) < 0)
+        return -1;
+
+    if (!pic32_pwm5_init(DEFAULT_PERIOD)
+    &&  !pic32_pwm5_set_duty_cycle(DEFAULT_PERIOD/2))
+        return 0;
+
+    return -1;
+}
+
+int pwm_enable(uint8_t mikrobus_index)
+{
+    if (check_valid_mikrobus(mikrobus_index) < 0)
+    {
+        fprintf(stderr, "pwm: Invalid mikrobus index.\n");
+        return -1;
+    }
+
+    return pic32_pwm5_enable();
+}
+
+int pwm_set_duty_cycle(uint8_t mikrobus_index, float percentage)
+{
+    if (check_valid_mikrobus(mikrobus_index) < 0)
+    {
+        fprintf(stderr, "pwm: Invalid mikrobus index.\n");
+        return -1;
+    }
+
+    if (percentage < 0.f || percentage > 100.f) {
+        fprintf(stderr, "pwm: Invalid percentage (must be in range 0..100).\n");
+        return -1;
+    }
+
+    pic32_pwm5_set_duty_cycle(pic32_pwm5_get_maximum_duty_cycle() * (percentage/100.f));
+
+    return 0;
+}
+
+int pwm_get_duty_cycle(uint8_t mikrobus_index, float *percentage)
+{
+    uint32_t max_duty_cycle, duty_cycle;
+
+    if (check_valid_mikrobus(mikrobus_index) < 0)
+    {
+        fprintf(stderr, "pwm: Invalid mikrobus index\n");
+        return -1;
+    }
+    if (percentage == NULL)
+    {
+        fprintf(stderr, "pwm: percentage cannot be null\n");
+        return -1;
+    }
+
+    max_duty_cycle = pic32_pwm5_get_maximum_duty_cycle();
+    if(max_duty_cycle == 0)
+        *percentage = 0.f;
+    else
+    {
+        duty_cycle = pic32_pwm5_get_duty_cycle() * 100;
+        *percentage = (float)duty_cycle / (float)max_duty_cycle;
+    }
+
+    return 0;
+}
+
+int pwm_set_frequency(uint8_t mikrobus_index, uint32_t frequency)
+{
+    return pwm_set_period(mikrobus_index, (1000 * 1000 * 1000) / frequency);
```

Comment:
It is only used twice. The name of the define would be quite confusing since it is used once to convert a frequency to a period and the second time from a period to a frequency.